src/lib/optimizer/strategy/chunk_pruning_rule.cpp

```diff
@@ -239,4 +242,107 @@ std::shared_ptr<TableStatistics> ChunkPruningRule::_prune_table_statistics(const
+     *  (2) check whether the predicate chain continues or ends with current_node.
+     */
+    auto predicate_chain_continues = true;
+    switch (current_node->type) {
+      // clang-format off
+      case LQPNodeType::Alias:
+      case LQPNodeType::Sort:
+      case LQPNodeType::StoredTable:
+      case LQPNodeType::Validate:
+      case LQPNodeType::Projection:   break;
+      case LQPNodeType::Predicate: {
+        const auto& predicate_node = std::static_pointer_cast<PredicateNode>(current_node);
+
+        // PredicateNode might not belong to the current predicate chain,
+        // e.g. when it follows a JoinNode and contains LQPColumnExpressions from other StoredTableNodes.
+        auto belongs_to_predicate_chain = true;
+        const auto& predicate_expression = predicate_node->predicate();
+        visit_expression(predicate_expression, [&](const auto& expression) {
+          if (expression->type != ExpressionType::LQPColumn) return ExpressionVisitation::VisitArguments;
+          const auto& column_expression = std::static_pointer_cast<LQPColumnExpression>(expression);
+          if (column_expression->original_node.lock() != stored_table_node) {
+            // PredicateNode filters stored_table_node. Therefore, we want to add it to the current predicate chain.
+            belongs_to_predicate_chain = false;
+          }
+          return ExpressionVisitation::DoNotVisitArguments;
+        });
+        if (belongs_to_predicate_chain) current_predicate_chain.emplace_back(predicate_node);
+      } break;
+      case LQPNodeType::Join: {
+        auto join_node = std::static_pointer_cast<JoinNode>(current_node);
+        if (join_node->join_mode != JoinMode::Semi) break;
+
+        // Check whether the predicate chain continues after the semi join
+        predicate_chain_continues = false;
+        for (const auto& expression : join_node->output_expressions()) {
+          if (expression->type != ExpressionType::LQPColumn) continue;
+          const auto column_expression = std::static_pointer_cast<LQPColumnExpression>(expression);
+          if (column_expression->original_node.lock() == stored_table_node) {
+            // At least one column expression of stored_table_node survives the semi join.
+            // Therefore, the predicate chain might continue.
+            predicate_chain_continues = true;
+            break;
+          }
+        }
+      } break;
+      default:
+        // For all other types of nodes, we finalize the predicate chain.
+        predicate_chain_continues = false;
+        // clang-format on
+    }
+
+    if (!predicate_chain_continues) {
+      // We do not have to go the LQP-tree up further because the predicate chain is complete.
+      predicate_chains.emplace_back(current_predicate_chain);
+      return LQPUpwardVisitation::DoNotVisitOutputs;
+    }
+
+    // Use recursion, in case the predicate chain branches.
```

Comment:
The whole point of this PR is that recursion when traversing the LQP is bad. How do we ensure that we do not visit nodes twice?

Reply:
- [ ] Add comment

Reply:
Add: "This is not an unnecessary duplication because..."?

Reply:
> Add: "This is not an unnecessary duplication because..."?

I added a comment to distinct the ChunkPruningRule from the BetweenCompositionRule at the top of between_composition_rule.cpp.

Or did you mean something else with `unnecessary duplication`? @mrks

Reply:
Looks good.